File: backend/listings/serializers/listing.py
from rest_framework import serializers
from ..models import Listing
from . import PhotoSerializer, LocationSerializer

from rest_framework_gis.fields import GeometryField


class ListingSerializer(serializers.ModelSerializer):
    """
    Main serializer for Listing model.
    Includes nested photos and locations.
    """

    photos = PhotoSerializer(many=True, read_only=True)
    locations = LocationSerializer(many=True, read_only=True)

    primary_location = serializers.SerializerMethodField()
    photo_count = serializers.SerializerMethodField()

    class Meta:
        model = Listing
        fields = (
            'id',
            'type',
            'status',
            'title',
            'description',
            'breed',
            'size',
            'color',
            'gender',
            'has_collar',
            'collar_color',
            'dog_name',
            'age_estimate',
            'special_marks',
            'fostering_address',
            'reward_offered',
            'search_radius_km',
            'photos',
            'photo_count',
            'locations',
            'primary_location',
            'created_at',
            'updated_at',
        )
        read_only_fields = (
            'id',
            'created_at',
            'updated_at',
        )

    def get_primary_location(self, obj):
        """
        Returns the primary (most recent) location for this listing.
        """
        primary = obj.locations.filter(is_primary=True).first()
        if primary:
            return LocationSerializer(primary).data
        return None

    def get_photo_count(self, obj):
        """
        Returns the total number of photos for this listing.
        """
        return obj.photos.count()

    def validate_search_radius_km(self, value):
        """
        Custom validation for search radius.
        For Lublin area, we limit radius to 1-10 km.
        """
        if value < 1 or value > 10:
            raise serializers.ValidationError(
                "Search radius must be between 1 and 10 km for Lublin area"
            )
        return value


class SimilarListingSerializer(serializers.Serializer):
    """
    Serializer for duplicate detection request.
    """
    search_type = serializers.ChoiceField(
        choices=['found', 'lost'],
        help_text="Type of listings to search: 'found' or 'lost'"
    )
    point = GeometryField(
        help_text="Geographic location (GeoJSON Point)"
    )
    breed = serializers.CharField(
        max_length=100,
        help_text="Dog breed"
    )
    size = serializers.ChoiceField(
        choices=['small', 'medium', 'large'],
        help_text="Dog size"
    )
    has_collar = serializers.BooleanField(
        help_text="Whether dog has a collar"
    )
    collar_color = serializers.CharField(
        max_length=50,
        required=False,
        allow_blank=True,
        help_text="Collar color (only if has_collar=True)"
    )

    def validate(self, data):
        """
        Check that collar_color is provided if has_collar=True
        """
        if data['has_collar'] and not data.get('collar_color'):
            raise serializers.ValidationError({
                'collar_color': 'Collar color is required when has_collar=True'
            })
        return data

File: mobile/types/listing.ts
export interface UserPreview {
	id: number;
	username: string;
	profile_photo: string | null;
}

export interface LocationPoint {
  type: 'Point';
  coordinates: [number, number]; // long, lat
}

export interface ListingLocation {
  id: number;
  added_by_user: number;
  listing: number;
  latitude: string;
  longitude: string;
  point: LocationPoint;
  address: string | null;
  location_type: 'exact' | 'approximate';
  is_primary: boolean;
  notes: string | null;
  accuracy_meters: number;
  created_at: string;
}

export interface ListingPhoto {
  id: number;
  cloudinary_url: string;
  thumbnail_url: string;
  order_index: number;
  uploaded_at: string;
}


export type ListingType = 'lost' | 'found'
type ListingStatus = 'active' | 'expired' | 'found' | 'returned'

export interface ListingItem {
  id: number;
  user: UserPreview;
  type: ListingType;
  status: ListingStatus;
  title: string;
  description: string;
  breed: string | null;
  size: string | null;
  color: string | null;
  gender: 'male' | 'female' | 'unknown';
  has_collar: boolean;
  collar_color: string | null;
  dog_name: string | null;
  age_estimate: string | null;
  special_marks: string | null;
  fostering_address: string | null;
  reward_offered: string | null;
  search_radius_km: number;
  photos: ListingPhoto[];
  photo_count: number;
  locations: ListingLocation[];
  primary_location: ListingLocation | null;
  created_at: string;
  updated_at: string;
}

